Unnamed arguments to `uses_labels()` check if label is set

rossellhayes opened the pull request:
You can now use unnamed arguments to `uses_labels()` to check if a a label for that aesthetic has been set.

``` r
library(ggcheck)
library(ggplot2)

p <- ggplot(data = mpg, mapping = aes(x = displ, y = hwy)) +
  geom_point(mapping = aes(color = class, shape = drv)) +
  labs(x = "Weight", y = "MPG", color = NULL)

# Unnamed arguments check if a label is set for the given aesthetic
uses_labels(p, "x", "title")
#>     x title 
#>  TRUE FALSE

# The check will return TRUE for values set to NULL
uses_labels(p, "color")
#> color 
#>  TRUE

# The check will return TRUE if an aesthetic has a default label
uses_labels(p, "shape")
#> shape 
#>  TRUE

# You can check if a label has the default value with default_label()
uses_labels(p, color = default_label(), shape = default_label())
#> color shape 
#> FALSE  TRUE

# You can also use default_label() outside uses_labels() if you add arguments
default_label(p, c("x", "y"))
#> $x
#> [1] "displ"
#> 
#> $y
#> [1] "hwy"
```

<sup>Created on 2021-11-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Closes #24.

rossellhayes commented:
@gadenbuie Sorry for the confusion, the functions do indeed work as you described. I accidentally copied a reprex from when I was still testing. I've updated with a reprex that reflects actual functionality.

For your more substantive question: is there a difference between `uses_labels(p, color = NA)` and `uses_aesthetics(p, "color")`?

gadenbuie commented:
>  is there a difference between `uses_labels(p, color = NA)` and `uses_aesthetics(p, "color")`?

Probably not, but I wouldn't want to recommend that authors know that. Internally, if we have `color = NA` we'd just skip the default value logic. That gives us some room for changes down the road, for example if we realize there are other ways of setting default values for non-aesthetic labels or something.

rossellhayes commented:
That makes sense. I'm not sure what the best interface would be though.

Unfortunately `NA` is a valid label value:

``` r
library(ggplot2)

p <- ggplot(data = mpg, mapping = aes(x = displ, y = hwy)) +
    geom_point() +
    labs(x = "Weight", y = NA)

p$labels
#> $x
#> [1] "Weight"
#> 
#> $y
#> [1] NA
is.na(p$labels$y)
#> [1] TRUE

p
```

![](https://i.imgur.com/zI2g2wu.png)

<sup>Created on 2021-11-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

I like the idea of having a `default_value()` function, but I don't think `uses_labels(p, x = default_value())` makes sense as an interface for checking if the label for `x` is set to _anything_. That makes it sound like we're testing `p$labels$x == default_value()`, not `exists(p$labels$x)`.

gadenbuie commented:
Good point, what about `default_label()` and `any_label()`?

```r
# check if color is *the default label*
uses_labels(p, color = default_label())
#> color
#> TRUE

# check if shape is *anything*
uses_labels(p, shape = any_label())
#> shape
#> TRUE

uses_labels(p, title = any_label())
#> title
#> FALSE
```

gadenbuie commented:
The alternative to this would be to have `uses_labels(p, "color")` return `TRUE` if the color label is set, regardless of it being a default value and then grading authors would use other methods to figure out if the label is the default.

I'm starting to think this might be the right way to go since it would simplify things considerably. We could still have `default_label()`, but it would take a plot and label as arguments, e.g. `default_label(p, "color")` and then this could be used to find out if the default is used

```r
uses_labels(p, color = default_label(p, "color"))

# or...
identical(get_labels(p, "color")$color, default_label(p, "color"))
```

rossellhayes commented:
Okay, that makes sense. Then authors could write code along the lines of
*`pass_if(uses_labels(p, "color") && !uses_labels(p, color = default_label()))`.

I'll work on adding the `default_label()` function and revert unnamed arguments to check if _any_ label is set.

gadenbuie commented:
Yeah and I'd recommend that `default_label()` has two modes (you can use S3 classes to dispatch on the first argument):

```r
# create a default label sentinel value
default_label()

# produces the default label for "color" based on the plot `p`
default_label(p, "color")
```

and then internally in `uses_labels()`, we'd convert the first case to the second.

rossellhayes commented:
I've added `default_label()` and updated the reprex to reflect current behavior.

`uses_label(p, "x")` now matches _any_ label for `x`, whether it is the default, `NULL`, or a new label set by the user. 

`uses_label(p, x = default_label())` checks if `x` has the default label.

brendanhcullen reviewed (approved):
@rossellhayes This looks great, and this new functionality will be really useful -- thanks! My only minor suggestion would be to consider renaming `default_label()` to `default_labels()` to be consistent with the plural form of other ggcheck functions (especially `uses_labels()`) that have the capacity to check multiple elements at once

rossellhayes commented:
@brendanhcullen Thank you for the feedback! I think keeping `default_label()` in the singular makes sense, because it's most common use will be in something like `uses_labels(p, x = default_label())`. It supports multiple arguments because there's no real reason it can't, but I think it will be primarily used in a context where the singular makes more sense.

brendanhcullen commented:
@rossellhayes Ah, great point. Agreed! 🥳

rossellhayes commented on R/labels.R:
```diff
@@ -76,6 +80,21 @@ get_labels <- function(p, aes = NULL) {
+#'   [Character][character] strings.
+#'
+#'   Unnamed arguments will check whether a label was set for that aesthetic.
+#'   This returns `TRUE` if the label exists and is not the default value.
```
Update this to reflect current behavior.

gadenbuie commented on R/labels.R:
```diff
@@ -87,28 +102,141 @@ uses_labels <- function(p, ...) {
+    return(logical(0))
+  }
+
+  args                  <- as.character(args)
```
```suggestion
  args <- as.character(args)
```

gadenbuie commented on R/labels.R:
```diff
@@ -76,6 +80,17 @@ get_labels <- function(p, aes = NULL) {
+#' @param p A ggplot object
+#' @param ... <[`dynamic-dots`][rlang::dyn-dots]>
+#'   [Character][character] strings.
+#'   Unnamed arguments will check whether a label exists that aesthetic.
```
```suggestion
#'   Unnamed arguments will check whether a label exists for that aesthetic.
```

gadenbuie commented on R/labels.R:
```diff
@@ -87,28 +102,141 @@ uses_labels <- function(p, ...) {
 
-  labels <- get_labels(p, names(args))
+  result         <- logical(length(args))
+  result[!named] <- check_labels_set(p, args[!named])
```
`args` make sense in this function as an object name since we've collected named and unnamed things from `...`, but once they're passed to `check_labels_set()` or `check_labels_match()` I think we need a better name for the object in those functions.

```r
# instead of
check_labels_set <- function(p, args) { ... }

# something like
check_labels_set <- function(p, labels) { ... }

# and/or
check_labels_match <- function(p, label_values) { ... }
```

rossellhayes replied:
Resolved in 3b9a20f

gadenbuie commented on R/labels.R:
```diff
@@ -87,28 +102,141 @@ uses_labels <- function(p, ...) {
+  )
+}
+
+check_labels_set <- function(p, args) {
```
```suggestion
check_labels_set <- function(p, labels) {
```

rossellhayes replied:
Resolved in 3b9a20f

gadenbuie commented on R/labels.R:
```diff
@@ -87,28 +102,141 @@ uses_labels <- function(p, ...) {
+  args %in% names(p$labels)
+}
+
+check_labels_match <- function(p, args) {
```
```suggestion
check_labels_match <- function(p, label_values) {
```

rossellhayes replied:
Resolved in 3b9a20f

gadenbuie commented on R/labels.R:
```diff
@@ -87,28 +102,141 @@ uses_labels <- function(p, ...) {
   result[null_expected]  <- lengths(labels[null_expected]) == 0
   result[!null_expected] <- purrr::map2_lgl(
-    args[!null_expected], labels[!null_expected], identical
+    args[!null_expected], labels[!null_expected],
+    ~ isTRUE(all.equal(.x, .y, check.attributes = FALSE, check.names = FALSE))
```
```suggestion
    args[!null_expected], 
    labels[!null_expected],
    function(expected, actual) {
      isTRUE(all.equal(expected, actual, check.attributes = FALSE, check.names = FALSE))
    }
```

gadenbuie commented on R/labels.R:
```diff
@@ -87,28 +102,141 @@ uses_labels <- function(p, ...) {
+
+  purrr::map(
+    aes,
+    function(aes) {
+      if (!is.null(p$mapping[[aes]])) {
+        return(as.character(make_labels(p$mapping[aes])))
+      }
+
+      for (i in seq_along(p$layers)) {
+        if (!is.null(p$layers[[i]]$mapping[[aes]])) {
+          return(as.character(make_labels(p$layers[[i]]$mapping[aes])))
+        }
+      }
+
+      NULL
+    }
```
If this function replicates behavior in ggplot2, it'd be nice to have a link to the source code to reference in case implementations diverge.

rossellhayes replied:
This is more of a reverse engineering of how `ggplot` works rather than a reimplementation of any of `ggplot`'s code, but 70ea762 adds comments explaining the logic of what's going on

gadenbuie replied:
The added comments are perfect, thanks! Future versions of ourselves will greatly appreciate them 😄

gadenbuie commented on R/labels.R:
```diff
@@ -87,28 +102,141 @@ uses_labels <- function(p, ...) {
+#' default_label(p, "color")
+#' default_label(p, "colour")
+#' @param p A [ggplot][ggplot2::ggplot] object
+#' @param aes If `aes` is a [character] vector, returns only the labels
+#'   corresponding to the included aesthetics.
+#'   Defaults to [`NULL`], which returns all labels.
```
```suggestion
#' @param aes If `aes` is a [character] vector, returns only the default labels
#'   (based on the plot `p`) that correspond to the included aesthetics.
#'   Defaults to [`NULL`], which returns the default values of all labels.
```

gadenbuie commented on R/labels.R:
```diff
@@ -87,28 +102,141 @@ uses_labels <- function(p, ...) {
+#'   corresponding to the included aesthetics.
+#'   Defaults to [`NULL`], which returns all labels.
+#'
+#' @return A named [list] in which each element is a [character] string
+#'   or [`NULL`]
```
It'd be better to document why/when the return value is character or NULL and how to interpret those values. You could easily pull from the comments in the examples

rossellhayes replied:
Resolved in e856502

gadenbuie commented on R/labels.R:
```diff
@@ -45,29 +45,33 @@ get_labels <- function(p, aes = NULL) {
-#'
-#' @return A logical vector of the same length as the number of inputs to `...`.
-#'
 #' @family functions for checking labels
 #' @export
```
Can we move these two lines to the very bottom of the documentation, just before the function definition?

rossellhayes replied:
Resolved in https://github.com/rstudio/ggcheck/pull/27/commits/a6c9305d17a1d2bfba7828de8e1622f935189a3a

gadenbuie commented on R/labels.R:
```diff
@@ -76,6 +80,17 @@ get_labels <- function(p, aes = NULL) {
 #' label_list <- list(x = "Weight", y = "MPG", color = NULL)
 #' uses_labels(p, label_list)
 #' uses_labels(p, !!!label_list)
+#' @param p A ggplot object
```
```suggestion
#'
#' @param p A ggplot object
```

rossellhayes replied:
This is the work of the automated GitHub Actions, so we should probably leave this until it won't be immediately reverted

gadenbuie commented on R/labels.R:
```diff
@@ -87,28 +102,141 @@ uses_labels <- function(p, ...) {
+#' # The colo(u)r aesthetic can be matched with or without a u
+#' default_label(p, "color")
+#' default_label(p, "colour")
+#' @param p A [ggplot][ggplot2::ggplot] object
```
```suggestion
#'
#' @param p A [ggplot][ggplot2::ggplot] object
```

gadenbuie reviewed (approved):
Looks great! All of my comments are fairly minor.

rossellhayes commented:
@gadenbuie I'm now seeing CI failures for `pkgdown`:
```
Cannot install packages:
* local::.: Can't install dependency rstudio/quillt@tidytemplate
* rstudio/quillt@tidytemplate: Can't find reference @tidytemplate in GitHub repo rstudio/quillt.
```
(https://github.com/rstudio/ggcheck/runs/4489235333)

rossellhayes commented on R/labels.R:
```diff
@@ -87,29 +101,150 @@ uses_labels <- function(p, ...) {
+  result         <- logical(length(args))
+  result[!named] <- check_labels_set(p, args[!named])
+  result[named]  <- check_labels_match(p, args[named])
+  names(result)  <- coalesce_chr(names(args), args)
```
Add comment about `colo(u)r`

gadenbuie commented on R/labels.R:
```diff
@@ -87,29 +101,150 @@ uses_labels <- function(p, ...) {
+    return(logical(0))
+  }
+
+  plot_labels <- get_labels(p, names(label_values))
```
Let's change `get_labels()` to ensure it always returns a named list

```r
list(a = 1)[c("a", "b")]
#> $a
#> [1] 1
#> 
#> $<NA>
#> NULL
```
